Realm/Realm/Handles/ObjectHandle.cs

```diff
@@ -191,9 +191,18 @@ public void SetValue(string propertyName, Metadata metadata, in RealmValue value
                             throw new RealmException("Can't link to an embedded object that is already managed.");
                         }
 
+                        if (GetProperty(propertyName, metadata).Type.IsRealmValue())
+                        {
+                            throw new NotSupportedException("Embedded objects cannot be used as a RealmValue.");
+                        }
+
                         var embeddedHandle = CreateEmbeddedObjectForProperty(propertyName, metadata);
                         realm.ManageEmbedded(embeddedObj, embeddedHandle);
                         return;
+
+                    // Asymmetric objects will not reach this path unless they are used as a RealmValue.
```

Comment:
I think it's worth spending a few more words on why this is the case. I assumed it's because of the weaver not compiling if `AsymmetricObject`s are linked to. I tried to verify that but of course there are other checks later in the process that at runtime throw if `AsymmetricObject`s are linked to.
So we could either write that it's the case because logically `AsymmetricObject`s can't be linked to, hence all the checks in place to keep this rule. Or one could point to where this comes from (weaver first, but even if not for that schema validation during the opening of the Realm).
And the same applies to the other comment on `AddToRealmIfNecessary`. We should say that it's the case because
1. the caller always filters out `EmbeddedObjects` 
2. and `AsymmetricObject`s for the same reason as this one

@nirinchev and @papafe opinions on this?

Reply:
Yes the Weaver will warn about the asym. object being linked to. It will hit line 386 in `RealmWeaver.cs`:
```c#
private WeavePropertyResult WeaveProperty(PropertyDefinition prop, TypeDefinition type)
{
    // ...

    if (prop.ContainsAsymmetricObject(_references))
    {
        return WeavePropertyResult.Warning($"{type.Name}.{prop.Name} is not an automatic property but its type is a AsymmetricObject. This usually indicates a relationship but AsymmetricObjects are not allowed to be the receiving end of any relationships.");
    }

    // ...
}
```

I believe there will also be a `RealmSchemaValidationException` when opening the realm.

For embedded objects that are not contained within a `RealmValue`, when being added to a collection, each individual method (e.g. `RealmList.Add()` and `RealmList.Insert()`) will only call `RealmCollectionBase.AddToRealmIfNecessary()` if it is not an embedded object. (Apparently with the exception of `RealmSet.Add`.)

Example from `RealmList.cs`:
```c#
public void Add(T value)
{
    var realmValue = Operator.Convert<T, RealmValue>(value);

    if (_isEmbedded)
    {
        // ...

        Realm.ManageEmbedded(EnsureUnmanagedEmbedded(realmValue), _listHandle.AddEmbedded());
        return;
    }

    AddToRealmIfNecessary(realmValue);
    _listHandle.Add(realmValue);
}
```

Comment update suggestion:
```c#
// Embedded and asymmetric objects will not reach this path unless the user explicitly adds
// them to the collection as RealmValues (e.g. IList<RealmValue>).
// Why will this path not be reached:
// * Embedded objects that are not contained within a RealmValue will be added to the collection
//   handle (e.g. _listHandle.AddEmbedded()) in the respective method (e.g. in RealmList.Add(),
//   RealmList.Insert(), RealmDictionary.Add()) rather than being added to the realm via
//   RealmCollectionBase.AddToRealmIfNecessary.
// * Asymmetric objects that are not contained within a RealmValue will produce a warning from
//   RealmWeaver.WeaveProperty() due to illegally linking to the object.
```

Reply:
Just as reference, if we assumed that the weaver was allowing links to `AsymmetricObject`s, you'd get the following from CORE:
```
Message: 
Realms.Exceptions.RealmException : A system error occurred while operating on a Realm. See InnerException for more details.
  ----> Realms.Exceptions.RealmSchemaValidationException : Schema validation failed due to the following errors:
- Property 'YourRealmObject.YourAsymmetricObjectProperty' of type 'object' cannot be a link to an asymmetric object.
```

Reply:
I tried to make it a little more concise and corrected a few details.

```
// Embedded and asymmetric objects can't reach this path unless the user explicitly adds
// them to the collection as RealmValues (e.g. IList<RealmValue>).
// This is because:
// * Plain embedded objects, beside RealmSet, are added by each collection
//   handle (e.g. _listHandle.AddEmbedded()) in the respective method (e.g. in RealmList.Add(),
//   RealmList.Insert(), RealmDictionary.Add(), etc.) rather than reaching
//   RealmCollectionBase.AddToRealmIfNecessary().
// * For plain asymmetric objects the weaver raises a compilation error since asymmetric
//   objects can't be linked to.
```

@papafe feedback?